plugin/evm/vm_test.go

```diff
@@ -1578,21 +1585,142 @@ func TestWaitForEvent(t *testing.T) {
+				require.NoError(t, blk.Accept(t.Context()))
+				require.NoError(t, blk2.Accept(t.Context()))
+			},
+		},
+		{
+			name: "WaitForEvent doesn't return once a block is built and accepted",
+			testCase: func(t *testing.T, vm *VM) {
+				t.Parallel()
 				ctx, cancel := context.WithTimeout(t.Context(), time.Millisecond*100)
 				defer cancel()
 
 				msg, err := vm.WaitForEvent(ctx)
 				require.ErrorIs(t, err, context.DeadlineExceeded)
 				require.Zero(t, msg)
+				var wg sync.WaitGroup
+				wg.Add(1)
+
+				signedTx := newSignedLegacyTx(t, vm.chainConfig, vmtest.TestKeys[0].ToECDSA(), 0, &vmtest.TestEthAddrs[1], big.NewInt(1), 21000, vmtest.InitialBaseFee, nil)
+
+				err = errors.Join(vm.txPool.AddRemotesSync([]*types.Transaction{signedTx})...)
+				require.NoError(t, err)
+
+				blk, err := vm.BuildBlock(t.Context())
+				require.NoError(t, err)
+
+				require.NoError(t, blk.Verify(t.Context()))
+
+				require.NoError(t, vm.SetPreference(t.Context(), blk.ID()))
+
+				require.NoError(t, blk.Accept(t.Context()))
+
+				ctx, cancel = context.WithTimeout(t.Context(), time.Millisecond*100)
+				defer cancel()
+
+				// We run WaitForEvent in a goroutine to ensure it can be safely called concurrently.
+				go func() {
+					defer wg.Done()
+					msg, err := vm.WaitForEvent(ctx)
+					require.ErrorIs(t, err, context.DeadlineExceeded)
+					require.Zero(t, msg)
+				}()
+
+				wg.Wait()
+			},
+		},
+		{
+			name: "WaitForEvent for two accepted blocks in a row",
+			testCase: func(t *testing.T, vm *VM) {
+				t.Parallel()
+				signedTx := newSignedLegacyTx(t, vm.chainConfig, vmtest.TestKeys[0].ToECDSA(), 0, &vmtest.TestEthAddrs[1], big.NewInt(1), 21000, vmtest.InitialBaseFee, nil)
+				err := errors.Join(vm.txPool.AddRemotesSync([]*types.Transaction{signedTx})...)
+				require.NoError(t, err)
+
+				blk, err := vm.BuildBlock(t.Context())
+				require.NoError(t, err)
+
+				require.NoError(t, blk.Verify(t.Context()))
+				require.NoError(t, vm.SetPreference(t.Context(), blk.ID()))
+
+				signedTx = newSignedLegacyTx(t, vm.chainConfig, vmtest.TestKeys[0].ToECDSA(), 1, &vmtest.TestEthAddrs[2], big.NewInt(1), 21000, vmtest.InitialBaseFee, nil)
+				err = errors.Join(vm.txPool.AddRemotesSync([]*types.Transaction{signedTx})...)
+				require.NoError(t, err)
+
+				time.Sleep(time.Second * 2)
+				blk2, err := vm.BuildBlock(t.Context())
+				require.NoError(t, err)
+
+				require.NoError(t, blk2.Verify(t.Context()))
+
+				signedTx = newSignedLegacyTx(t, vm.chainConfig, vmtest.TestKeys[0].ToECDSA(), 2, &vmtest.TestEthAddrs[2], big.NewInt(1), 21000, vmtest.InitialBaseFee, nil)
+				err = errors.Join(vm.txPool.AddRemotesSync([]*types.Transaction{signedTx})...)
+				require.NoError(t, err)
+				// We run WaitForEvent in a goroutine to ensure it can be safely called concurrently.
+				var wg sync.WaitGroup
+
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					msg, err := vm.WaitForEvent(t.Context())
+					require.Equal(t, commonEng.PendingTxs, msg)
```

Comment:
I think we determined that `require` should not be used in a goroutine because `t.Fatal` exits the function by throwing a panic. Which will result in the test crashing rather than failing normally. Can we use the same pattern as in `WaitForEvent returns when a transaction is added to the mempool` to ensure the concurrent behavior of `WaitForEvent` works as expected?

We can probably move
```golang
type result struct {
	msg commonEng.Message
	err error
}
```
out of the above subtest and into this parent test to avoid copying the definition.